tests/test_base.py

```diff
@@ -336,3 +336,211 @@ def test_ensure_text_inputs_errors():
+    grade = 1
+    for i in range(4, 7):
+        grade -= 0.3
+        expected_result = {'msg': '', 'grade_decimal': None, 'ok': 'partial'}
+        result = grader(None, 'cat', attempt=i)
+        assert result['msg'] == expected_result['msg']
+        assert result['ok'] == expected_result['ok']
+        assert result['grade_decimal'] == approx(grade)
+
+    for i in range(7, 10):
+        expected_result = {'msg': '', 'grade_decimal': 0.1, 'ok': 'partial'}
+        result = grader(None, 'cat', attempt=i)
+        assert result['msg'] == expected_result['msg']
+        assert result['ok'] == expected_result['ok']
+        assert result['grade_decimal'] == approx(0.1)
+
+    # Test individual flags
+    # Ensure it turns off properly
+    grader = StringGrader(
+        answers='cat',
+        attempt_based_credit=False
+    )
+
+    expected_result = {'msg': '', 'grade_decimal': 1, 'ok': True}
+    for i in range(1, 10):
+        assert grader(None, 'cat', attempt=i) == expected_result
+
+    # Ensure that wrong answers are always zeros
+    grader = StringGrader(
+        answers='cat',
+        attempt_based_credit=True
+    )
+
+    expected_result = {'msg': '', 'grade_decimal': 0, 'ok': False}
+    for i in range(1, 10):
+        assert grader(None, 'dog', attempt=i) == expected_result
+
+    # Ensure that zero credit is graded as false appropriately
+    grader = StringGrader(
+        answers='cat',
+        attempt_based_credit=True,
+        minimum_credit=0,
+        attempt_based_credit_msg=False
+    )
+
+    expected_result = {'msg': '', 'grade_decimal': 0, 'ok': False}
+    for i in range(10, 20):
+        assert grader(None, 'cat', attempt=i) == expected_result
+
+    # Ensure that messages are included as appropriate
+    grader = StringGrader(
+        answers='cat',
+        attempt_based_credit=True,
+        minimum_credit=0,
+        decrease_credit_after=1,
+        decrease_credit_steps=2,
+        attempt_based_credit_msg=True
+    )
+
+    expected_result = {'msg': '', 'grade_decimal': 1, 'ok': True}
+    assert grader(None, 'cat', attempt=1) == expected_result
+    expected_result = {'msg': '', 'grade_decimal': 0, 'ok': False}
+    assert grader(None, 'dog', attempt=1) == expected_result
+
+    expected_result = {'msg': 'Maximum credit for attempt #2 is 0.50.', 'grade_decimal': 0.5, 'ok': 'partial'}
+    assert grader(None, 'cat', attempt=2) == expected_result
+    expected_result = {'msg': '', 'grade_decimal': 0, 'ok': False}
+    assert grader(None, 'dog', attempt=2) == expected_result
+
+    expected_result = {'msg': 'Maximum credit for attempt #3 is 0.00.', 'grade_decimal': 0, 'ok': False}
+    assert grader(None, 'cat', attempt=3) == expected_result
+    expected_result = {'msg': '', 'grade_decimal': 0, 'ok': False}
+    assert grader(None, 'dog', attempt=3) == expected_result
+
+    # Ensure that messages are appended as appropriate
+    grader = StringGrader(
+        answers={'expect': 'cat', 'msg': 'Meow!'},
```

Comment:
Do you want to change this to `answers={'expect': 'cat', 'grade_decimal': 0.5, 'msg': 'Meow!'},` just to asssert somewhere that this does work as expected on partial-credit answers?

(It does work this way, just not asserted in a test unless I missed it.)

Reply:
Done.